File: cs/Problems/Day8.cs
namespace aoc24.Problems;

// https://adventofcode.com/2024/day/8
public sealed class Day8 : IProblem<int>
{
    public int Solve(string input) => CalculateAntiNodesOptimized(input);

    private readonly Dictionary<char, List<Point>> antennaGroups = [];

    private readonly HashSet<Point> antiNodes = [];

    public int CalculateAntiNodesOptimized(ReadOnlySpan<char> input)
    {
        Span<char> inner = stackalloc char[Matrix<char>.SizeFor(input)];
        var matrix = Matrix<char>.CreateFrom(input, inner);

        antennaGroups.Clear();
        antiNodes.Clear();

        // Group antennas by character.
        for (int y = 0; y < matrix.Height; y++)
        {
            for (int x = 0; x < matrix.Width; x++)
            {
                var location = new Point(x, y);

                if (matrix.ItemAt(location) is char val and not '.')
                {
                    if (!antennaGroups.ContainsKey(val))
                        antennaGroups[val] = [];

                    antennaGroups[val].Add(location);
                }
            }
        }

        // Compare antennas and calculate the anti-nodes in each group.
        foreach (var (_, antennas) in antennaGroups)
        {
            for (int i = 0; i < antennas.Count - 1; i++)
            {
                for (int j = i + 1; j < antennas.Count; j++)
                {
                    var (antennaA, antennaB) = (antennas[i], antennas[j]);
                    var distance = antennaA - antennaB;

                    var antiNodeA = antennaA + distance;
                    var antiNodeB = antennaB - distance;

                    if (matrix.ItemAt(antiNodeA) is not null)
                        antiNodes.Add(antiNodeA);

                    if (matrix.ItemAt(antiNodeB) is not null)
                        antiNodes.Add(antiNodeB);
                }
            }
        }

        return antiNodes.Count;
    }
}

File: go/problems/day8.go
package problems

import (
	"image"
	"strings"

	"github.com/Meryovi/hobby-aoc-2024/go/shared"
)

type Day8 struct{}

func (d Day8) Solve(input string) int {
	matrix := strings.Split((input), shared.NewLine)
	my := len(matrix)
	mx := len(matrix[0])

	groups := make(map[byte][]image.Point)
	antiNodes := make(map[image.Point]image.Point)

	for y := range my {
		for x := range mx {
			c := matrix[y][x]
			if c != '.' {
				pt := image.Pt(x, y)
				groups[c] = append(groups[c], pt)
			}
		}
	}

	for _, antennas := range groups {
		for i := range len(antennas) - 1 {
			for j := i + 1; j < len(antennas); j++ {
				a := antennas[i]
				b := antennas[j]
				d := a.Sub(b)

				antiA := image.Pt(a.X+d.X, a.Y+d.Y)
				antiB := image.Pt(b.X-d.X, b.Y-d.Y)

				if inBounds(antiA, mx, my) {
					antiNodes[antiA] = antiA
				}
				if inBounds(antiB, mx, my) {
					antiNodes[antiB] = antiB
				}
			}
		}
	}

	return len(antiNodes)
}

func inBounds(pt image.Point, mx, my int) bool {
	return pt.X >= 0 && pt.X < mx && pt.Y >= 0 && pt.Y < my
}
